generator-test/src/test/java/com/linkedin/pegasus/generator/test/TestMaskMap.java

```diff
@@ -191,4 +193,145 @@ public void testNonAliasedUnionFieldNestedProjection()
     ));
     Assert.assertEquals(mask.getDataMap(), tree.getDataMap());
   }
+
+  // Tests the case where a partial mask is  created and updated later.
+  @Test
+  public void testReuseMaskSimpleFields()
+  {
+    RecordTest.ProjectionMask mask = RecordTest.createMask()
+        .withBooleanField();
+
+    MaskTree tree = MaskCreator.createPositiveMask(PathSpecSet.of(
+        RecordTest.fields().booleanField()));
+    Assert.assertEquals(mask.getDataMap(), tree.getDataMap());
+
+    // Now update the mask to add new fields.
+    mask.withDoubleField();
+
+    MaskTree tree2 = MaskCreator.createPositiveMask(PathSpecSet.of(
+        RecordTest.fields().booleanField(),
+        RecordTest.fields().doubleField()));
+    Assert.assertEquals(mask.getDataMap(), tree2.getDataMap());
+  }
+
+  @Test
+  public void testReuseMaskNestedRecord()
+  {
+    RecordTest.ProjectionMask mask = RecordTest.createMask()
+        .withBooleanField()
+        .withRecordField(nestedMask -> nestedMask.withLocation());
+
+    MaskTree tree = MaskCreator.createPositiveMask(PathSpecSet.of(
+        RecordTest.fields().booleanField(),
+        RecordTest.fields().recordField().location()));
+    Assert.assertEquals(mask.getDataMap(), tree.getDataMap());
+
+    // Now update the mask to add new fields.
+    mask.withDoubleField()
+        .withRecordField(nestedMask -> nestedMask.withOptionalLocation());
+
+    MaskTree tree2 = MaskCreator.createPositiveMask(PathSpecSet.of(
+        RecordTest.fields().booleanField(),
+        RecordTest.fields().doubleField(),
+        RecordTest.fields().recordField().location(),
+        RecordTest.fields().recordField().optionalLocation()));
+    Assert.assertEquals(mask.getDataMap(), tree2.getDataMap());
+  }
+
+  @Test
+  public void testReuseMaskNestedRecordClearing()
+  {
+    RecordTest.ProjectionMask mask = RecordTest.createMask()
+        .withRecordField(nestedMask -> nestedMask.withLocation());
+
+    MaskTree tree = MaskCreator.createPositiveMask(PathSpecSet.of(
+        RecordTest.fields().recordField().location()));
+    Assert.assertEquals(mask.getDataMap(), tree.getDataMap());
+
+    // Clear the nested mask by projecting the entire field.
+    mask.withRecordField();
+    tree = MaskCreator.createPositiveMask(PathSpecSet.of(
+        RecordTest.fields().recordField()));
```

Comment:
I guess I am learning here, so there is a field called `location` inside the nested `recordField` in the `RecordTest` and it was added in the pathSpec. Now we clear that `location` in the PathSpec by adding all the other fields in the nested `recordField`?

Reply:
Correct. When the entire record is projected, it includes all fields, which means you don't have to explicitly project the "location" (2nd level nested) field anymore.